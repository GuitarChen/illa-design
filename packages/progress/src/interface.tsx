import { HTMLAttributes, ReactNode } from "react"

export type ProgressColorScheme =
  | string
  | "white"
  | "blackAlpha"
  | "gray"
  | "grayBlue"
  | "red"
  | "orange"
  | "yellow"
  | "green"
  | "blue"
  | "cyan"
  | "purple"
  | "techPink"
  | "techPurple"

export type ProgressType = "line" | "circle" | "miniCircle" | "miniRing"

export type ProgressStatus = "normal" | "success" | "error"

export type ProgressSize = "small" | "medium" | "large"

export interface ProgressProps extends HTMLAttributes<HTMLDivElement> {
  type?: ProgressType
  steps?: number
  status?: ProgressStatus
  color?: ProgressColorScheme
  trailColor?: ProgressColorScheme
  showText?: boolean
  formatText?: (percent: number) => ReactNode
  percent?: number
  width?: string
  strokeWidth?: string
}
