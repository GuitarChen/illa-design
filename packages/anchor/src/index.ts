export * from "./anchor"
export * from "./interface"
