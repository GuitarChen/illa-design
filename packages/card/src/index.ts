export * from "./card"
export * from "./interface"
export * from "./card-grid"
export * from "./meta"
