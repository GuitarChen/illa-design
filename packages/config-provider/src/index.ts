export * from "./config-provider"
export * from "./interface"
export * from "./config-provider-context"
export * from "./config"

// locale
export * from "./locale/en-US"
export * from "./locale/zh-CN"
export * from "./locale/def"
