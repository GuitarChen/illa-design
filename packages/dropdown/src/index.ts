export * from "./dropdown"
export * from "./interface"
