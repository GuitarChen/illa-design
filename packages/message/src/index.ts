export * from "./style"
export * from "./message"
export * from "./interface"
