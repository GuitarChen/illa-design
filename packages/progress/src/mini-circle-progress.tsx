import { forwardRef, ReactNode } from "react"
import { ProgressProps } from "./interface"
import { SuccessIcon, WarningCircleIcon } from "@illa-design/icon"
import { Trigger } from "@illa-design/trigger"
import { applyContainer } from "./common-style"
import {
  applyCircleProgressContainer,
  applyCircleStatus,
  applyCircleSvgContainer,
} from "./mini-circle-progress-style"
import { globalColor, illaPrefix } from "@illa-design/theme"

export const MiniCircleProgress = forwardRef<HTMLDivElement, ProgressProps>(
  (props, ref) => {
    const {
      type = "line",
      status = "normal",
      color = "blue",
      width = "16px",
      strokeWidth = "4px",
      trailColor = "gray",
      showText = true,
      formatText = (percent: number) => {
        return `${percent}%`
      },
      percent = 0,
      steps,
      ...otherProps
    } = props

    const radius = `calc(${width} / 2)`
    const center = `calc(${width} / 2)`

    let finalColor: string

    switch (status) {
      case "normal":
        finalColor = color
        break
      case "success":
        finalColor = "green"
        break
      case "error":
        finalColor = "red"
        break
    }

    let statusComponent: ReactNode
    switch (status) {
      case "success":
        statusComponent = (
          <SuccessIcon color={globalColor(`--${illaPrefix}-green-03`)} />
        )
        break
      case "error":
        statusComponent = (
          <WarningCircleIcon color={globalColor(`--${illaPrefix}-red-03`)} />
        )
        break
    }

    return (
      <div css={applyContainer(width)} ref={ref} {...otherProps}>
        <Trigger disabled={!showText} content={formatText(percent)}>
          {(status == "normal" || status == "error") && (
            <svg css={applyCircleSvgContainer(width, trailColor)}>
              <circle
                css={applyCircleProgressContainer(finalColor, width, percent)}
                fill="none"
                cx={center}
                cy={center}
                r={radius}
                strokeWidth={width}
              />
            </svg>
          )}
          {status == "success" && (
            <span css={applyCircleStatus}>{statusComponent}</span>
          )}
        </Trigger>
      </div>
    )
  },
)

MiniCircleProgress.displayName = "MiniCircleProgress"
