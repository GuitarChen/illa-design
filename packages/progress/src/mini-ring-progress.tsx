import { forwardRef } from "react"
import { ProgressProps } from "./interface"
import { Trigger } from "@illa-design/trigger"
import {
  applyContainer,
  applyProgressContainer,
  applyProgressContainerBg,
  applySvgContainer,
} from "./common-style"

export const MiniRingProgress = forwardRef<HTMLDivElement, ProgressProps>(
  (props, ref) => {
    const {
      type = "line",
      steps,
      status = "normal",
      color = "blue",
      width = "16px",
      strokeWidth = "4px",
      trailColor = "gray",
      showText = true,
      formatText = (percent: number) => {
        return `${percent}%`
      },
      percent = 0,
      ...otherProps
    } = props

    const radius = `calc((${width} - ${strokeWidth}) / 2)`
    const center = `calc(${width} / 2)`

    let finalColor: string

    switch (status) {
      case "normal":
        finalColor = color
        break
      case "success":
        finalColor = "green"
        break
      case "error":
        finalColor = "red"
        break
    }

    return (
      <div css={applyContainer(width)} ref={ref} {...otherProps}>
        <Trigger disabled={!showText} content={formatText(percent)}>
          <svg css={applySvgContainer(width)}>
            <circle
              css={applyProgressContainerBg(trailColor, width, strokeWidth)}
              fill="none"
              cx={center}
              cy={center}
              r={radius}
              strokeWidth={strokeWidth}
            />
            <circle
              css={applyProgressContainer(
                finalColor,
                width,
                strokeWidth,
                percent,
              )}
              fill="none"
              cx={center}
              cy={center}
              r={radius}
              strokeWidth={strokeWidth}
            />
          </svg>
        </Trigger>
      </div>
    )
  },
)

MiniRingProgress.displayName = "MiniRingProgress"
