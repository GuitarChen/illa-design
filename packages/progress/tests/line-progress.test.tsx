import { render, screen } from "@testing-library/react"
import "@testing-library/jest-dom"
import { Progress } from "../src"

test("Progress renders correctly", () => {
  render(
    <div>
      <Progress data-testid="test-progress" type="line" />
    </div>,
  )
  expect(screen.getByTestId("test-progress")).toBeInTheDocument()
})

test("Progress renders with colors", () => {
  render(
    <div>
      <Progress
        data-testid="test-progress-red"
        type="line"
        color="red"
        trailColor="red"
      />
      <Progress
        data-testid="test-progress-gray"
        type="line"
        color="gray"
        trailColor="gray"
      />
      <Progress
        data-testid="test-progress-custom"
        type="line"
        color="#123123"
        trailColor="#321321"
      />
      <Progress
        data-testid="test-progress-red-step"
        type="line"
        steps={3}
        color="red"
        trailColor="red"
      />
      <Progress
        data-testid="test-progress-grayBlue-step"
        type="line"
        steps={3}
        color="gray"
        trailColor="gray"
      />
      <Progress
        data-testid="test-progress-custom-step"
        type="line"
        steps={3}
        color="#123123"
        trailColor="#321321"
      />
    </div>,
  )
  expect(screen.getByTestId("test-progress-red")).toMatchSnapshot()
  expect(screen.getByTestId("test-progress-gray")).toMatchSnapshot()
  expect(screen.getByTestId("test-progress-custom")).toMatchSnapshot()
  expect(screen.getByTestId("test-progress-red-step")).toMatchSnapshot()
  expect(screen.getByTestId("test-progress-grayBlue-step")).toMatchSnapshot()
  expect(screen.getByTestId("test-progress-custom-step")).toMatchSnapshot()
})

test("Progress renders with different status", () => {
  render(
    <div>
      <Progress status="success" type="line" />,
      <Progress status="error" type="line" />,
    </div>,
  )
  expect(screen.getByTitle("SuccessIcon")).toBeInTheDocument()
  expect(screen.getByTitle("WarningCircleIcon")).toBeInTheDocument()
})

test("Progress renders without text", () => {
  render(<Progress percent={50} showText={false} type="line" />)
  expect(screen.queryByText("50")).not.toBeTruthy()
})

test("Progress renders with format text function", () => {
  render(
    <Progress
      percent={50}
      type="line"
      formatText={(percent) => {
        return `test${percent}`
      }}
    />,
  )
  expect(screen.getByText("test50")).toBeInTheDocument()
})

test("Progress renders with custom width", () => {
  render(
    <Progress
      data-testid="test-progress"
      percent={50}
      width="100px"
      type="line"
      strokeWidth="10px"
    />,
  )
  expect(screen.getByTestId("test-progress")).toMatchSnapshot()
})

test("Progress renders with steps", () => {
  render(
    <Progress data-testid="test-progress" steps={3} type="line" percent={40} />,
  )
  expect(screen.getByTestId("test-progress")).toMatchSnapshot()
})
