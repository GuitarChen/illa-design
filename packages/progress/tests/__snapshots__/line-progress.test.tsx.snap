// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress renders with colors 1`] = `
.emotion-0 {
  vertical-align: middle;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.emotion-1 {
  display: inline-block;
  height: 4px;
  width: 320px;
  position: relative;
}

.emotion-2 {
  position: absolute;
  background-color: #fcc;
  width: 100%;
  border-radius: calc(4px / 2);
  height: 4px;
}

.emotion-3 {
  -webkit-transition: width 200ms;
  transition: width 200ms;
  position: absolute;
  width: 0%;
  border-radius: calc(4px / 2);
  background-color: #ff4747;
  height: 4px;
}

.emotion-4 {
  margin-left: 8px;
  line-height: 14px;
  font-size: 14px;
  color: #a9aeb8;
}

<div
  class="emotion-0"
  data-testid="test-progress-red"
>
  <div
    class="emotion-1"
  >
    <div
      class="emotion-2"
    />
    <div
      class="emotion-3"
    />
  </div>
  <span
    class="emotion-4"
  >
    0%
  </span>
</div>
`;

exports[`Progress renders with colors 2`] = `
.emotion-0 {
  vertical-align: middle;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.emotion-1 {
  display: inline-block;
  height: 4px;
  width: 320px;
  position: relative;
}

.emotion-2 {
  position: absolute;
  background-color: #ebebeb;
  width: 100%;
  border-radius: calc(4px / 2);
  height: 4px;
}

.emotion-3 {
  -webkit-transition: width 200ms;
  transition: width 200ms;
  position: absolute;
  width: 0%;
  border-radius: calc(4px / 2);
  background-color: #1f1f1f;
  height: 4px;
}

.emotion-4 {
  margin-left: 8px;
  line-height: 14px;
  font-size: 14px;
  color: #a9aeb8;
}

<div
  class="emotion-0"
  data-testid="test-progress-gray"
>
  <div
    class="emotion-1"
  >
    <div
      class="emotion-2"
    />
    <div
      class="emotion-3"
    />
  </div>
  <span
    class="emotion-4"
  >
    0%
  </span>
</div>
`;

exports[`Progress renders with colors 3`] = `
.emotion-0 {
  vertical-align: middle;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.emotion-1 {
  display: inline-block;
  height: 4px;
  width: 320px;
  position: relative;
}

.emotion-2 {
  position: absolute;
  background-color: #321321;
  width: 100%;
  border-radius: calc(4px / 2);
  height: 4px;
}

.emotion-3 {
  -webkit-transition: width 200ms;
  transition: width 200ms;
  position: absolute;
  width: 0%;
  border-radius: calc(4px / 2);
  background-color: #123123;
  height: 4px;
}

.emotion-4 {
  margin-left: 8px;
  line-height: 14px;
  font-size: 14px;
  color: #a9aeb8;
}

<div
  class="emotion-0"
  data-testid="test-progress-custom"
>
  <div
    class="emotion-1"
  >
    <div
      class="emotion-2"
    />
    <div
      class="emotion-3"
    />
  </div>
  <span
    class="emotion-4"
  >
    0%
  </span>
</div>
`;

exports[`Progress renders with colors 4`] = `
.emotion-0 {
  vertical-align: middle;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.emotion-1 {
  display: inline-block;
  height: 4px;
  width: 320px;
  position: relative;
}

.emotion-2 {
  width: 100%;
  position: absolute;
}

.emotion-3 {
  display: inline-flex;
  vertical-align: middle;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.emotion-4 {
  height: 100%;
  display: inline-flex;
  margin-right: 4px;
  margin-bottom: 0px;
}

.emotion-5 {
  background-color: #fcc;
  width: calc((320px - (3 - 1) * 4px) / 3);
  border-radius: calc(4px / 2);
  height: 4px;
}

.emotion-8 {
  height: 100%;
  display: inline-flex;
  margin-right: 0px;
  margin-bottom: 0px;
}

.emotion-13 {
  width: 0px;
  border-radius: calc(4px / 2);
  background-color: #ff4747;
  height: 4px;
}

.emotion-18 {
  margin-left: 8px;
  line-height: 14px;
  font-size: 14px;
  color: #a9aeb8;
}

<div
  class="emotion-0"
  data-testid="test-progress-red-step"
>
  <div
    class="emotion-1"
  >
    <div
      class="emotion-2 emotion-3"
    >
      <div
        class="emotion-4"
      >
        <div
          class="emotion-5"
        />
      </div>
      <div
        class="emotion-4"
      >
        <div
          class="emotion-5"
        />
      </div>
      <div
        class="emotion-8"
      >
        <div
          class="emotion-5"
        />
      </div>
    </div>
    <div
      class="emotion-2 emotion-3"
    >
      <div
        class="emotion-4"
      >
        <div
          class="emotion-13"
        />
      </div>
      <div
        class="emotion-4"
      >
        <div
          class="emotion-13"
        />
      </div>
      <div
        class="emotion-8"
      >
        <div
          class="emotion-13"
        />
      </div>
    </div>
  </div>
  <span
    class="emotion-18"
  >
    0%
  </span>
</div>
`;

exports[`Progress renders with colors 5`] = `
.emotion-0 {
  vertical-align: middle;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.emotion-1 {
  display: inline-block;
  height: 4px;
  width: 320px;
  position: relative;
}

.emotion-2 {
  width: 100%;
  position: absolute;
}

.emotion-3 {
  display: inline-flex;
  vertical-align: middle;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.emotion-4 {
  height: 100%;
  display: inline-flex;
  margin-right: 4px;
  margin-bottom: 0px;
}

.emotion-5 {
  background-color: #ebebeb;
  width: calc((320px - (3 - 1) * 4px) / 3);
  border-radius: calc(4px / 2);
  height: 4px;
}

.emotion-8 {
  height: 100%;
  display: inline-flex;
  margin-right: 0px;
  margin-bottom: 0px;
}

.emotion-13 {
  width: 0px;
  border-radius: calc(4px / 2);
  background-color: #1f1f1f;
  height: 4px;
}

.emotion-18 {
  margin-left: 8px;
  line-height: 14px;
  font-size: 14px;
  color: #a9aeb8;
}

<div
  class="emotion-0"
  data-testid="test-progress-grayBlue-step"
>
  <div
    class="emotion-1"
  >
    <div
      class="emotion-2 emotion-3"
    >
      <div
        class="emotion-4"
      >
        <div
          class="emotion-5"
        />
      </div>
      <div
        class="emotion-4"
      >
        <div
          class="emotion-5"
        />
      </div>
      <div
        class="emotion-8"
      >
        <div
          class="emotion-5"
        />
      </div>
    </div>
    <div
      class="emotion-2 emotion-3"
    >
      <div
        class="emotion-4"
      >
        <div
          class="emotion-13"
        />
      </div>
      <div
        class="emotion-4"
      >
        <div
          class="emotion-13"
        />
      </div>
      <div
        class="emotion-8"
      >
        <div
          class="emotion-13"
        />
      </div>
    </div>
  </div>
  <span
    class="emotion-18"
  >
    0%
  </span>
</div>
`;

exports[`Progress renders with colors 6`] = `
.emotion-0 {
  vertical-align: middle;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.emotion-1 {
  display: inline-block;
  height: 4px;
  width: 320px;
  position: relative;
}

.emotion-2 {
  width: 100%;
  position: absolute;
}

.emotion-3 {
  display: inline-flex;
  vertical-align: middle;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.emotion-4 {
  height: 100%;
  display: inline-flex;
  margin-right: 4px;
  margin-bottom: 0px;
}

.emotion-5 {
  background-color: #321321;
  width: calc((320px - (3 - 1) * 4px) / 3);
  border-radius: calc(4px / 2);
  height: 4px;
}

.emotion-8 {
  height: 100%;
  display: inline-flex;
  margin-right: 0px;
  margin-bottom: 0px;
}

.emotion-13 {
  width: 0px;
  border-radius: calc(4px / 2);
  background-color: #123123;
  height: 4px;
}

.emotion-18 {
  margin-left: 8px;
  line-height: 14px;
  font-size: 14px;
  color: #a9aeb8;
}

<div
  class="emotion-0"
  data-testid="test-progress-custom-step"
>
  <div
    class="emotion-1"
  >
    <div
      class="emotion-2 emotion-3"
    >
      <div
        class="emotion-4"
      >
        <div
          class="emotion-5"
        />
      </div>
      <div
        class="emotion-4"
      >
        <div
          class="emotion-5"
        />
      </div>
      <div
        class="emotion-8"
      >
        <div
          class="emotion-5"
        />
      </div>
    </div>
    <div
      class="emotion-2 emotion-3"
    >
      <div
        class="emotion-4"
      >
        <div
          class="emotion-13"
        />
      </div>
      <div
        class="emotion-4"
      >
        <div
          class="emotion-13"
        />
      </div>
      <div
        class="emotion-8"
      >
        <div
          class="emotion-13"
        />
      </div>
    </div>
  </div>
  <span
    class="emotion-18"
  >
    0%
  </span>
</div>
`;

exports[`Progress renders with custom width 1`] = `
.emotion-0 {
  vertical-align: middle;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.emotion-1 {
  display: inline-block;
  height: 10px;
  width: 100px;
  position: relative;
}

.emotion-2 {
  position: absolute;
  background-color: #ebebeb;
  width: 100%;
  border-radius: calc(10px / 2);
  height: 10px;
}

.emotion-3 {
  -webkit-transition: width 200ms;
  transition: width 200ms;
  position: absolute;
  width: 50%;
  border-radius: calc(10px / 2);
  background-color: #1e6fff;
  height: 10px;
}

.emotion-4 {
  margin-left: 8px;
  line-height: 14px;
  font-size: 14px;
  color: #a9aeb8;
}

<div
  class="emotion-0"
  data-testid="test-progress"
>
  <div
    class="emotion-1"
  >
    <div
      class="emotion-2"
    />
    <div
      class="emotion-3"
    />
  </div>
  <span
    class="emotion-4"
  >
    50%
  </span>
</div>
`;

exports[`Progress renders with steps 1`] = `
.emotion-0 {
  vertical-align: middle;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.emotion-1 {
  display: inline-block;
  height: 4px;
  width: 320px;
  position: relative;
}

.emotion-2 {
  width: 100%;
  position: absolute;
}

.emotion-3 {
  display: inline-flex;
  vertical-align: middle;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-flex-wrap: nowrap;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.emotion-4 {
  height: 100%;
  display: inline-flex;
  margin-right: 4px;
  margin-bottom: 0px;
}

.emotion-5 {
  background-color: #ebebeb;
  width: calc((320px - (3 - 1) * 4px) / 3);
  border-radius: calc(4px / 2);
  height: 4px;
}

.emotion-8 {
  height: 100%;
  display: inline-flex;
  margin-right: 0px;
  margin-bottom: 0px;
}

.emotion-13 {
  width: calc((320px - (3 - 1) * 4px) / 3);
  border-radius: calc(4px / 2);
  background-color: #1e6fff;
  height: 4px;
}

.emotion-15 {
  width: 0px;
  border-radius: calc(4px / 2);
  background-color: #1e6fff;
  height: 4px;
}

.emotion-18 {
  margin-left: 8px;
  line-height: 14px;
  font-size: 14px;
  color: #a9aeb8;
}

<div
  class="emotion-0"
  data-testid="test-progress"
>
  <div
    class="emotion-1"
  >
    <div
      class="emotion-2 emotion-3"
    >
      <div
        class="emotion-4"
      >
        <div
          class="emotion-5"
        />
      </div>
      <div
        class="emotion-4"
      >
        <div
          class="emotion-5"
        />
      </div>
      <div
        class="emotion-8"
      >
        <div
          class="emotion-5"
        />
      </div>
    </div>
    <div
      class="emotion-2 emotion-3"
    >
      <div
        class="emotion-4"
      >
        <div
          class="emotion-13"
        />
      </div>
      <div
        class="emotion-4"
      >
        <div
          class="emotion-15"
        />
      </div>
      <div
        class="emotion-8"
      >
        <div
          class="emotion-15"
        />
      </div>
    </div>
  </div>
  <span
    class="emotion-18"
  >
    40%
  </span>
</div>
`;
